File: adapter/src/main/java/com/sovegetables/adapter/DiffCallBack.java
package com.sovegetables.adapter;


import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

/**
 * Created by albert on 2018/1/26.
 * 一个通用DiffUtil.Callback，继承DiffUtil.Callback.
 * 数据实体类需要继承DiffCallBack.DiffContent接口。
 */

public class DiffCallBack<T extends DiffCallBack.DiffContent<T>> extends DiffUtil.Callback{

    private List<T> mOldData;
    private List<T> mNewData;

    public DiffCallBack(List<T> oldData, List<T> newData) {
        mOldData = oldData;
        mNewData = newData;
    }

    @Override
    public int getOldListSize() {
        return mOldData == null? 0 : mOldData.size();
    }

    @Override
    public int getNewListSize() {
        return mNewData == null? 0 : mNewData.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        try {
            return mNewData.get(newItemPosition).areItemsTheSame(mOldData.get(oldItemPosition));
        }catch (Exception e){
            e.printStackTrace();
        }
        return false;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        try {
            return mNewData.get(newItemPosition).areContentsTheSame(mOldData.get(oldItemPosition));
        }catch (Exception e){
            e.printStackTrace();
        }
        return false;
    }

    public interface DiffContent<T>{
        boolean areItemsTheSame(T t);
        boolean areContentsTheSame(T t);
    }
}

File: adapter/src/main/java/com/sovegetables/adapter/DiffCallBack.kt
package com.sovegetables.adapter

import androidx.recyclerview.widget.DiffUtil

class DiffCallBack<T : DiffCallBack.DiffItem<T>>(private val mOldData: List<T>?, private val mNewData: List<T>?) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return mOldData?.size ?: 0
    }

    override fun getNewListSize(): Int {
        return mNewData?.size ?: 0
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        try {
            return mNewData!![newItemPosition].areItemsTheSame(mOldData!![oldItemPosition])
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return false
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        try {
            return mNewData!![newItemPosition].areContentsTheSame(mOldData!![oldItemPosition])
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return false
    }

    interface DiffItem<T> {
        fun areItemsTheSame(t: T): Boolean
        fun areContentsTheSame(t: T): Boolean
    }
}